controlplane/kubeadm/internal/cluster_test.go

```diff
@@ -0,0 +1,117 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package internal
+
+import (
+	"context"
+	"testing"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha3"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func TestGetMachinesForCluster(t *testing.T) {
+	m := ManagementCluster{Client: &fakeClient{}}
+	clusterKey := types.NamespacedName{
+		Namespace: "my-namespace",
+		Name:      "my-cluster",
+	}
+	machines, err := m.GetMachinesForCluster(context.Background(), clusterKey)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if len(machines) != 3 {
+		t.Fatalf("expected 3 machines but found %d", len(machines))
+	}
+
+	// Test the OwnedControlPlaneMachines works
+	machines, err = m.GetMachinesForCluster(context.Background(), clusterKey, OwnedControlPlaneMachines("my-control-plane"))
+	if err != nil {
+		t.Fatal(err)
+	}
+	if len(machines) != 1 {
+		t.Fatalf("expected 1 control plane machine but got %d", len(machines))
+	}
+
+	// Test that filters AND the logic together
+	nameFilter := func(cluster clusterv1.Machine) bool {
+		return cluster.Name == "first-machine"
+	}
+	machines, err = m.GetMachinesForCluster(context.Background(), clusterKey, OwnedControlPlaneMachines("my-control-plane"), nameFilter)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if len(machines) != 1 {
+		t.Fatalf("expected 1 control plane machine but got %d", len(machines))
+	}
+}
+
+type fakeClient struct {
```

Comment:
The `fake.NewClient` is being deprecated and this is a more true unit test solution.

Reply:
I wouldn't worry too much about the deprecation until it happens, we're also getting more involved upstream, so it remains to be seen if it's going to go away or not